// mist-device/test/utils/ucp-interface.js


function DummyInterface() {
    // Options are used for adding context like request source
    this.opts = {};
    this.app = { config: { wsid: 'dummy-wsid' } };
}

DummyInterface.prototype.attachUcpDevice = function(device) {
    this.ucpDevice = device;
};

DummyInterface.prototype.write = function(feature, value, cb) {
    this.ucpDevice.write(feature, value, this.opts, cb);
};

DummyInterface.prototype.read = function(feature, cb) {
    this.ucpDevice.read(feature, this.opts, cb);
};

DummyInterface.prototype.invoke = function(feature, args, cb) {
    this.ucpDevice.invoke(feature, args, this.opts, cb);
};

DummyInterface.prototype.model = function(cb) {
    this.ucpDevice.model(this.opts, cb);
};

DummyInterface.prototype.subscribe = function(feature, endpoint, cb) {
    this.ucpDevice.subscribe(feature, endpoint, this.opts, cb);
};

DummyInterface.prototype.map = function(from, to, settings, opts, cb) {
    if (arguments.length === 4) {
        cb = opts;
        settings = {};
    }
    
    this.ucpDevice.map(from, to, settings, this.opts, cb);
};


module.exports = { 
    DummyInterface: DummyInterface };
