// wish-app/src/support.js
var fs = require('fs');
var path = require('path');

function createConfig(serviceConfig) {

    var file = serviceConfig.configFile;

    if ( process.platform === 'win32' ) {
        file = process.env.LOCALAPPDATA+"/Wish/"+file;
    } else {
        file = path.resolve(process.env.HOME+"/.wish/"+file);
    }

    var config = {
        service: { meta: serviceConfig, instance: serviceConfig.wsid ? serviceConfig.wsid : 'service.x' },
        ui: serviceConfig.ui,
        core: { 
            ws: serviceConfig.core.ws ? serviceConfig.core.ws : 'ws://localhost:9090',
            api: serviceConfig.core.api
        }
    };

    try {
        config = JSON.parse(fs.readFileSync(file));
    } catch (e) {
        try {
            fs.writeFileSync(file, JSON.stringify(config));
        } catch(e) {
            console.log("Could not read or write "+file+". Failed to start.");
            process.exit(1);
        }
    }
    
    config.core.api = serviceConfig.core.api;
    
    return config;
}

module.exports = {
    createConfig: createConfig };
